k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -105,270 +103,277 @@ class SingleNsNamer(
   }
 }
 
+object SingleNsNamer {
+  protected val PrefixLen = 2
+}
+
 abstract class EndpointsNamer(
   idPrefix: Path,
-  mkApi: String => NsApi,
-  backoff: Stream[Duration] = Backoff.exponentialJittered(10.milliseconds, 10.seconds)
-)(implicit timer: Timer = DefaultTimer) extends EnumeratingNamer {
+  mkApi: String => v1.NsApi,
+  labelName: Option[String] = None,
+  backoff: Stream[Duration] = EndpointsNamer.DefaultBackoff
+)(implicit timer: Timer = DefaultTimer)
+  extends Namer with Stabilize {
 
   import EndpointsNamer._
+  protected[this] val variablePrefixLength: Int
 
-  def lookup(path: Path): Activity[NameTree[Name]]
+  // memoize port remapping watch activities so that we don't have to
+  // create multiple watches on the same `Services` API object.
+  private[this] val portRemappings =
```

Comment:
I was wondering if something like that existed, but all I could find was stuff from scalaz which I figured you *probably* wouldn't approve of us depending on 🙃